### DISCUSSION TITLE
Android File System Question

### DISCUSSION THREAD

**Discussion by seanocali** [unoplatform/uno#9882] [Q&A]
How can I retrieve a StorageFile or StorageFolder that was chosen from a picker in Android in a previous session? 

In UWP I can save the path or I can add it to the FutureAccessList and save the key. But in Android if I use the picker and choose the Download folder, the path is `"/tree/primary:Download"`. How could I retrieve that as a StorageFolder or access it using System.IO methods?

**Comment by seanocali** [comment]
Maybe a better question:
Is there a way to parse `"/tree/primary:Download"` into an absolute path or a URI?

**Comment by workgroupengineering** [comment]
Hi, 
- to creaate file in StorageFolder you can use [CreateFileAsync](https://learn.microsoft.com/en-us/uwp/api/windows.storage.storagefolder.createfileasync?view=winrt-22621#windows-storage-storagefolder-createfileasync(system-string))
- to open file StorageFile.[OpenAsync
](https://learn.microsoft.com/en-us/uwp/api/windows.storage.storagefile.openasync?view=winrt-22621)

If to use Android primitives, there is a bug #9616 which I don't know if fixed in the Uno stable release.

In Uno dev release you can use like this:

```csharp
       Storage.IStorageFile file; 
       ...
	var localpath = file.Path;
	var fileDescriptor = localpath.StartsWith('/')
			? ParcelFileDescriptor.Open(new Java.IO.File(localpath), ParcelFileMode.ReadOnly)
			: Android.App.Application.Context.ContentResolver.OpenFileDescriptor(Android.Net.Uri.Parse(localpath), "r");
``

**Reply by seanocali** [reply]
Appreciate the response but not what I'm looking for. Creating, opening, or getting a fileDescriptor of a known file isn't the issue here.

**Reply by workgroupengineering** [reply]
> Appreciate the response but not what I'm looking for. Creating, opening, or getting a fileDescriptor of a known file isn't the issue here.

file is return of the picker.



if you don't need to use Android primitives, use this as suggested above

```csharp
using System.IO;
....

FileOpenPicker openPicker = new FileOpenPicker();
openPicker.ViewMode = PickerViewMode.Thumbnail;
openPicker.SuggestedStartLocation = PickerLocationId.PicturesLibrary;
openPicker.FileTypeFilter.Add(".jpg");
openPicker.FileTypeFilter.Add(".jpeg");
openPicker.FileTypeFilter.Add(".png");

StorageFile file = await openPicker.PickSingleFileAsync();
if (file != null)
{
    // Application now has read/write access to the picked file
    OutputTextBlock.Text = "Picked photo: " + file.Name;
    var stream = (await file.OpenAsync(FileAccessMode.Read,StorageOpenOptions.None)).AsStream();
}
else
{
    OutputTextBlock.Text = "Operation cancelled.";
}

```

**Reply by seanocali** [reply]
I'm sorry if the question was unclear. I'm not asking how to pick a file. I'm asking about persistence. After a user has picked a folder or file location, how can that location be stored and later retrieved in another runtime session without the user having to use the picker again?

**Reply by pkar70** [reply]
Just "by the way": SuggestedStartLocation is unsupported on Android, although it isn't marked as unsupported. It was supported in #2896 (unmerged), but about year later Android API was changed and obsoletes this PR.

**Comment by seanocali** [comment]
I'll try be more clear. Here is a simple use-case. Let's say an app asks a user to tell it where their tax documents are stored. The user chooses a folder using a folder picker. The path to that location is stored persistently, and the next time the application is loaded it can retrieve that folder without the user having to use the picker again. It can check if that folder still exists and it has access to it, and it can then scan for documents. 

Pretty standard stuff applications do all the time. It's basically like "Open Recent" in Microsoft Word. Yet I can't seem to do this on Android with Uno.

In UWP when the user chooses a folder using a picker the StorageFolder.Path is the absolute path to that folder. But in Android using Uno this is not the case. Instead it's a relative path with a scheme like "/tree/primary:Download" or "tree/[deviceName]:Documents". I have a WPF/UWP background not Android, so as far as I can tell these relative paths are useless. They can't be used to create a StorageFolder object, and they can't be used to create a DirectoryInfo object. I need either an absolute path or a URI.

Now with simple string split and replace I can parse "/tree/primary:Download" into "/storage/emulated/0/Download" which works on my Android emulator, but is that always going to work on all Android devices?  Could the mount point name be different in different Android versions?
